hazelcast/src/main/java/com/hazelcast/map/impl/SizeEstimator.java

```
@@ -16,18 +16,101 @@
+
+    public abstract long getCost(T record);
+
+    public abstract void reset();
+
+    private static boolean hasCompressedOops() {
+        String enabled = isHotSpotCompressedOopsEnabledOrNull();
+
+        if (enabled != null) {
+            return Boolean.valueOf(enabled);
+        }
+
+        try {
+            // fallback for other VM implementations.
+            return isCompressedOopsEnabled();
+        } catch (Exception e) {
+            LOGGER.warning(e);
+        }
+
+        // by default act as compressed oops are enabled.
+        return true;
+    }
+
+    private static String isHotSpotCompressedOopsEnabledOrNull() {
+        try {
+            MBeanServer server = ManagementFactory.getPlatformMBeanServer();
+            ObjectName mbean = new ObjectName("com.sun.management:type=HotSpotDiagnostic");
+            Object[] objects = {"UseCompressedOops"};
+            String[] strings = {"java.lang.String"};
+            String operation = "getVMOption";
+            CompositeDataSupport compressedOopsValue = (CompositeDataSupport) server.invoke(mbean, operation, objects, strings);
+            return compressedOopsValue.get("value").toString();
+        } catch (Exception e) {
+            LOGGER.warning("Failed to read HotSpot specific configuration", e);
+        }
+
+        return null;
+    }
 
-    long getSize();
 
-    void add(long size);
+    /*
+     * Fallback when checking CompressedOopsEnabled.
+     *
+     * Borrowed from http://openjdk.java.net/projects/code-tools/jol/
+     */
+    private static boolean isCompressedOopsEnabled() {
+        // When running with CompressedOops on 64-bit platform, the address size
+        // reported by Unsafe is still 8, while the real reference fields are 4 bytes long.
+        // Try to guess the reference field size with this naive trick.
+        int oopSize;
+        try {
+            long off1 = UNSAFE.objectFieldOffset(TestCompressedOops.class.getField("obj1"));
```

You need to check if `Unsafe` is available and NOT fail when not available.